presto-verifier/src/main/java/com/facebook/presto/verifier/framework/PrestoAction.java

```diff
@@ -0,0 +1,247 @@
+    {
+        R apply(ResultSet resultSet)
+                throws SQLException;
+
+        ResultSetConverter<List<Object>> DEFAULT = resultSet -> {
+            ImmutableList.Builder<Object> row = ImmutableList.builder();
+            for (int i = 0; i < resultSet.getMetaData().getColumnCount(); i++) {
+                row.add(resultSet.getObject(i));
+            }
+            return row.build();
+        };
+    }
+
+    private static final String QUERY_MAX_EXECUTION_TIME = "query_max_execution_time";
+
+    private final SqlExceptionClassifier exceptionClassifier;
+
+    private final Map<QueryGroup, String> clusterUrls;
+    private final Duration controlTimeout;
+    private final Duration testTimeout;
+    private final Duration metadataTimeout;
+    private final Duration checksumTimeout;
+
+    private final RetryDriver networkRetry;
+    private final RetryDriver prestoRetry;
+
+    @Inject
+    public PrestoAction(
+            SqlExceptionClassifier exceptionClassifier,
+            VerifierConfig config,
+            @ForClusterConnection RetryConfig networkRetryConfig,
+            @ForPresto RetryConfig prestoRetryConfig)
+    {
+        this.exceptionClassifier = requireNonNull(exceptionClassifier, "exceptionClassifier is null");
+        this.clusterUrls = ImmutableMap.of(
+                CONTROL, config.getControlJdbcUrl(),
+                TEST, config.getTestJdbcUrl());
+        this.controlTimeout = requireNonNull(config.getControlTimeout(), "controlTimeout is null");
+        this.testTimeout = requireNonNull(config.getTestTimeout(), "testTimeout is null");
+        this.metadataTimeout = requireNonNull(config.getMetadataTimeout(), "metadataTimeout is null");
+        this.checksumTimeout = requireNonNull(config.getChecksumTimeout(), "checksumTimeout is null");
+
+        this.networkRetry = new RetryDriver(networkRetryConfig, queryException -> queryException.getType() == CLUSTER_CONNECTION);
+        this.prestoRetry = new RetryDriver(prestoRetryConfig, queryException -> queryException.getType() == PRESTO && queryException.isRetryable());
+    }
+
+    public QueryStats execute(
+            Statement statement,
+            QueryConfiguration configuration,
+            QueryOrigin queryOrigin)
+    {
+        return execute(statement, configuration, queryOrigin, Optional.empty()).getQueryStats();
+    }
+
+    public <R> QueryResult<R> execute(
+            Statement statement,
+            QueryConfiguration configuration,
+            QueryOrigin queryOrigin,
+            ResultSetConverter<R> converter)
+    {
+        return execute(statement, configuration, queryOrigin, Optional.of(converter));
+    }
+
+    private <R> QueryResult<R> execute(
+            Statement statement,
+            QueryConfiguration configuration,
+            QueryOrigin queryOrigin,
+            Optional<ResultSetConverter<R>> converter)
+    {
+        return prestoRetry.run(
+                "presto",
```

Comment:
Should there be a more meaningful name passed here? I am thinking about the following options (multiple can be included) - 

1) kind of retry - network or presto
2) some query specific identifier
3) stage of the query - setup, main, non-deterministic, teardown

Reply:
Add 1 and 3, not sure about what do you mean by "some query specific identifier"
I'm not too worry about this since it is for debugging logging only.

Reply:
> some query specific identifier

I meant something which identifies the specific query that we are testing so that if we see this error message, we can trace it back to a specific query. I did have a comment somewhere which talked about actually logging about the failures. These can be helpful during debugging. So something which we can do if we are trying to figure out what happened to a specific test case.